src/Common/GuardedPoolAllocatorCommon.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <base/defines.h>
+#include <Common/StackTrace.h>
+
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+
+#include <cstdio>
+
+namespace clickhouse_gwp_asan
+{
+
+struct ThreadLocalPackedVariables
+{
+    constexpr ThreadLocalPackedVariables() : next_sample_counter(0), recursive_guard(false) { }
+    uint32_t next_sample_counter = 0;
+    bool recursive_guard = false;
+};
+
+static_assert(sizeof(ThreadLocalPackedVariables) <= sizeof(uint64_t), "Thread-locals do not fit in one cache line (uint64_t)");
+
+enum class Error : uint8_t
+{
+    UNKNOWN,
+    USE_AFTER_FREE,
+    DOUBLE_FREE,
+    INVALID_FREE,
+    BUFFER_OVERFLOW,
+    BUFFER_UNDERFLOW
+};
+
+const char * errorToString(const Error & error);
+
+static constexpr uint64_t kInvalidThreadID = UINT64_MAX;
+
+struct GuardedPoolAllocatorState
```

Comment:
Also it will be great if you will add layout in the comment, that will include:
- user pointers
- guard pages for that user pointers
- metadata